cmd/ftl-lsp-gen/main.go

```diff
@@ -0,0 +1,172 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+type hover struct {
+	// Match this text for triggering this hover, e.g. "//ftl:typealias"
+	Match string `json:"match"`
+
+	// Source file to read from.
+	Source string `json:"source"`
+
+	// Select these heading to use for the docs. If omitted, the entire markdown file is used.
+	// Headings are included in the output.
+	Select []string `json:"select,omitempty"`
+}
+
+func main() {
+	hovers := parseHover()
+	collected := make(map[string]string, len(hovers))
+
+	for _, hover := range hovers {
+		path := filepath.Join("docs", "content", "docs", hover.Source)
+		file, err := os.Open(path)
+		if err != nil {
+			panic(err)
+		}
+
+		content, title := getMarkdownWithTitle(file)
+
+		if len(hover.Select) > 0 {
+			output := ""
+			for _, sel := range hover.Select {
+				output += selector(content, sel)
+			}
+			content = output
+		} else {
+			// We need to inject a heading for the hover content.
+			content = fmt.Sprintf("## %s\n\n%s", title, content)
+		}
+
+		collected[hover.Match] = content
+	}
+
+	writeGoFile(collected)
+}
+
+func parseHover() []hover {
+	path := "lsp/hover.json"
+
+	file, err := os.Open(path)
+	if err != nil {
+		panic(err)
```

Comment:
It's best not to panic even in CLI tools, instead return the error and use Kong to terminate with a useful message.